### PR TITLE
[SBOM] Keep layer info in SBOM components

### PR DESCRIPTION
**Pull request by lebauce**
### What does this PR do?

Change the sbom.container_image.overlayfs_direct_scan mode to
keep the layers information attached to each component of the SBOM.

This PR also bump Trivy and uses a custom walker to implement
most of the features previously implemented in our Trivy fork.
 
### Motivation

This allows to match vulnerabilities to the layer they are part of.

### Describe how you validated your changes

Set `sbom.container_image.overlayfs_direct_scan` in both host and containerized mode,
for the docker, containerd and crio runtimes.

### CODE REVIEW

**Review by chouetz — APPROVED** [review]

File: internal/tools/go.mod [hunk 1]
```diff
@@ -56,7 +56,7 @@ require (
 	github.com/chavacava/garif v0.1.0 // indirect
 	github.com/chigopher/pathlib v0.19.1 // indirect
 	github.com/ckaznocha/intrange v0.1.2 // indirect
-	github.com/cloudflare/circl v1.3.7 // indirect
+	github.com/cloudflare/circl v1.3.8 // indirect
```
Comment (pducolin) [inline internal/tools/go.mod:59]: Might conflict with https://github.com/DataDog/datadog-agent/pull/32452 who's touching the same dependency, hopefully github does merge correctly both

**Review by pducolin — APPROVED** [review]
ok for changes owned by @DataDog/agent-devx-loops

File: pkg/util/trivy/crio.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+}
+
+// ScanCRIOImageFromOverlayFS scans the CRI-O image layers using OverlayFS.
+func (c *Collector) ScanCRIOImageFromOverlayFS(ctx context.Context, imgMeta *workloadmeta.ContainerImageMetadata, client crio.Client, scanOptions sbom.ScanOptions) (sbom.Report, error) {
+	lowerDirs, err := client.GetCRIOImageLayers(imgMeta)
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve layer directories: %w", err)
+	}
+
+	var diffIDs []string
```
Comment (paulcacheux) [inline pkg/util/trivy/crio.go:77]: nit
```suggestion
	diffIDs := nake([]string, 0, len(lowerDirs)
```

File: pkg/util/trivy/docker.go [hunk 3]
```diff
@@ -52,9 +57,85 @@ func convertDockerImage(ctx context.Context, client client.ImageAPIClient, imgMe
...
 	}
 
-	return &image{
+	img := &image{
 		opener:  imageOpener(ctx, DockerCollector, imageID, f, client.ImageSave),
 		inspect: inspect,
 		history: configHistory(history),
-	}, cleanup, nil
+	}
+	img.Image = img
```
Comment (paulcacheux) [inline pkg/util/trivy/docker.go:65]: `img.Image == img` ? that feels strange

File: pkg/util/trivy/overlayfs.go [hunk 4]
```diff
@@ -3,240 +3,109 @@
 // This product includes software developed at Datadog (https://www.datadoghq.com/).
 // Copyright 2016-present Datadog, Inc.
 
-//go:build linux
+//go:build trivy
```
Comment (paulcacheux) [inline pkg/util/trivy/overlayfs.go:6]: maybe `trivy && linux` ?

File: pkg/util/trivy/crio.go [hunk 5]
```diff
@@ -0,0 +1,94 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build trivy
```
Comment (paulcacheux) [inline pkg/util/trivy/crio.go:6]: I think there is a build tag for crio

Comment (lebauce) [reply pkg/util/trivy/crio.go]: :+1:

File: pkg/util/trivy/docker.go (hunk 3 above)
Comment (lebauce) [reply pkg/util/trivy/docker.go]: I don't remember why. Removed it to try as it indeed looks strange

**Review by paulcacheux — APPROVED** [review]

**Review by sblumenthal — APPROVED** [review]

File: go.mod [hunk 6]
```diff
@@ -1062,23 +1082,13 @@ replace github.com/vishvananda/netlink => github.com/DataDog/netlink v1.0.1-0.20
 // Pull in replacements needed by upstream Trivy
 replace (
 	// Maps to Trivy fork https://github.com/DataDog/trivy/commits/use-fs-main-dd/
-	github.com/aquasecurity/trivy => github.com/DataDog/trivy v0.0.0-20241216135157-95e0e96002ee
+	github.com/aquasecurity/trivy => github.com/DataDog/trivy v0.0.0-20241223160824-d6db1f1a84ce
```
[the remark this answers is not in the document]
Comment (lebauce) [reply go.mod:1085]: The comment is indeed incorrect. The new branch seems to be based on `main` of Trivy and includes a commit (https://github.com/DataDog/trivy/commit/f11c14ff339283c0844b2e98b809d7824cc7d98f) that bundles a bundle of commits. I did my best to check to check that we are not missing something here.

File: pkg/sbom/scanner/scanner.go [hunk 7]
```diff
@@ -26,7 +26,7 @@ import (
 )
 
 const (
-	defaultScanTimeout = 30 * time.Second
+	defaultScanTimeout = 60 * time.Second
```
[the remark this answers is not in the document]
Comment (lebauce) [reply pkg/sbom/scanner/scanner.go:29]: Good catch, this was for debugging purposes

**Pull request merged**